Adding support for separate auth+capture on BNPLs

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #3086

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

This PR enables the separate auth/capture feature for the new BNPL payment methods. This is achieved by changing the return of `requires_automatic_capture()` to `true`. Also, some texts were updated to reflect this change.

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

- Checkout to this branch on your local environment (`add/bnpl-support-for-auth-capture`)
- Run `npm install`, `npm run build:webpack` and `npm run up`
- Connect your Stripe account credentials in settings (`http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`) and configure your webhooks
- Check the box "Issue an authorization on checkout, and capture later"
- To test the capture flow you must enable external access to your local environment (ngrok for example) or use a public store

### Affirm

- Set your store currency to USD or CAD. And your Stripe account country to US or CA, respectively
- Confirm that you can enable Affirm
- As a shopper, add at least 50 USD (or CAD) to your cart
- Complete the checkout
- Confirm that you can see the order as "On-hold" in wp-admin
<img width="407" alt="Screenshot 2024-05-06 at 15 57 12" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/cc2f2f06-9b48-4a5f-9d44-e39b9d5c77ed">

- Confirm that the order is displayed as "Uncaptured" on Stripe's dashboard
<img width="275" alt="Screenshot 2024-05-06 at 15 58 53" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/c468ff8b-7381-498f-b534-bc903bc23261">

- Capture the payment on the order details screen (Stripe dashboard)
- Confirm the order is now "Completed"

### Klarna

- Set your store currency to any supported (AUD, CAD, CHF, CZK, DKK, EUR, GBP, NOK, NZD, PLN, SEK, or USD). And your Stripe account country to match it.
- Confirm that you can enable Klarna
- As a shopper, add any product to your cart
- Complete the checkout
- Confirm that you can see the order as "On-hold" in wp-admin
<img width="407" alt="Screenshot 2024-05-06 at 15 57 12" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/cc2f2f06-9b48-4a5f-9d44-e39b9d5c77ed">

- Confirm that the order is displayed as "Uncaptured" on Stripe's dashboard
<img width="275" alt="Screenshot 2024-05-06 at 15 58 53" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/c468ff8b-7381-498f-b534-bc903bc23261">

- Capture the payment on the order details screen (Stripe dashboard)
- Confirm the order is now "Completed"

### Afterpay/Clearpay

- Set your store currency to any supported (USD, CAD, GBP, AUD, or NZD). And your Stripe account country to match it.
- Confirm that you can enable Afterpay/Clearpay
- As a shopper, add at least 1 USD (or any other currency) to your cart
- Complete the checkout
- Confirm that you can see the order as "On-hold" in wp-admin
<img width="407" alt="Screenshot 2024-05-06 at 15 57 12" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/cc2f2f06-9b48-4a5f-9d44-e39b9d5c77ed">

- Confirm that the order is displayed as "Uncaptured" on Stripe's dashboard
<img width="275" alt="Screenshot 2024-05-06 at 15 58 53" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/c468ff8b-7381-498f-b534-bc903bc23261">

- Capture the payment on the order details screen (Stripe dashboard)
- Confirm the order is now "Completed"

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

wjrosa committed:
Adding support for separate auth+capture on BNPLs

wjrosa committed:
Merge branch 'develop' into add/bnpl-support-for-auth-capture

wjrosa committed:
Allow toggling of the BNPL methods in the settings page

wjrosa committed:
Fix webhook handlers

Mayisha committed:
Merge branch 'develop' into add/bnpl-support-for-auth-capture

Mayisha reviewed (changes requested):
Hi @wjrosa. :wave: 

I found two issues while testing this PR.

1. Affirm is not offered on the checkout page with manual capture enabled (Klarna and Afterpay are there). As this PR https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3123 has been merged, we need to add the `requires_automatic_capture` again in the Affirm class to make it available. 

2. I could successfully place orders with Klarna and Afterpay. But I capture the payment on the order details screen (Stripe dashboard), the orders in my wp-admin is still in `on-hold` state. I found that I am receiving the webhook event properly but my payment methods ids are `stripe_klarna` and `stripe_afterpay_clearpay`. And so the `WC_Stripe_Helper::payment_method_allows_manual_capture( $order->get_payment_method() )` is returning flase.

wjrosa committed:
Adding support for manual capture to Affirm method

wjrosa committed:
Fix manual capture verification helper method

wjrosa committed:
Adding specific unit test for the new helper method

wjrosa commented:
> Hi @wjrosa. 👋
> 
> I found two issues while testing this PR.
> 
> 1. Affirm is not offered on the checkout page with manual capture enabled (Klarna and Afterpay are there). As this PR [Hide BNPLs on the checkout pages when capture later is enabled. #3123](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3123) has been merged, we need to add the `requires_automatic_capture` again in the Affirm class to make it available.
> 2. I could successfully place orders with Klarna and Afterpay. But I capture the payment on the order details screen (Stripe dashboard), the orders in my wp-admin is still in `on-hold` state. I found that I am receiving the webhook event properly but my payment methods ids are `stripe_klarna` and `stripe_afterpay_clearpay`. And so the `WC_Stripe_Helper::payment_method_allows_manual_capture( $order->get_payment_method() )` is returning flase.

Hi Mayisha! Thanks for your review!

1. The PR is now updated with the `develop` branch and I have re-included the `requires_automatic_capture` capability
2. Sorry about that! I could not do the whole test before my current rotation. I have updated the helper method and included a specific unit test for it

Mayisha committed:
Merge branch 'develop' into add/bnpl-support-for-auth-capture

Mayisha reviewed (approved):
Thanks @wjrosa. This is working great now. :tada: 

- I can enable/disable Affirm, Afterpay and Klarna from the settings page when manual capture is enabled.
- I can choose Affirm, Afterpay and Klarna on the checkout page.
- The order status is `on-hold` after checkout. (for all 3 methods)
- When I capture the payment from the Stripe dashboard, the order status is changed to `processing`.
